tool/tsh/common/putty_config_windows.go

```diff
@@ -0,0 +1,360 @@
+	// to typos or similar. setting an "invalid" key in the registry makes it impossible to delete via the PuTTY
+	// UI and requires registry edits, so it's much better to error out early here.
+	hostname := tc.Config.Host
+	if !puttyhosts.NaivelyValidateHostname(hostname) {
+		return trace.BadParameter("provided hostname %v does not look like a valid hostname. Make sure it doesn't contain illegal characters.", hostname)
+	}
+
+	port := tc.Config.HostPort
+	userHostString := hostname
+	login := ""
+	if tc.Config.HostLogin != "" {
+		login = tc.Config.HostLogin
+		userHostString = fmt.Sprintf("%v@%v", login, userHostString)
+	}
+
+	// connect to proxy to fetch cluster info
+	proxyClient, err := tc.ConnectToProxy(cf.Context)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer proxyClient.Close()
+
+	// parse out proxy details
+	proxyHost, _, err := net.SplitHostPort(tc.Config.SSHProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// get root cluster name and set keypaths
+	rootClusterName, err := proxyClient.RootClusterName(cf.Context)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	keysDir := profile.FullProfilePath(tc.Config.KeysDir)
+	ppkFilePath := keypaths.PPKFilePath(keysDir, proxyHost, tc.Config.Username)
+	certificateFilePath := keypaths.SSHCertPath(keysDir, proxyHost, tc.Config.Username, rootClusterName)
+
+	// if a leaf cluster is requested, we need to check whether we can find it while running the CA loop below
+	wantLeaf := cf.LeafClusterName != ""
+	leafExists := false
+
+	// get all CAs for the cluster (including trusted clusters)
+	var cas []types.CertAuthority
+	err = tc.WithRootClusterClient(cf.Context, func(clt auth.ClientI) error {
+		cas, err = clt.GetCertAuthorities(cf.Context, types.HostCA, false /* exportSecrets */)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	})
+	// iterate over all the CAs
+	hostCAPublicKeys := make(map[string][]string)
+	for _, ca := range cas {
+		// if this is either the root or the requested leaf cluster, process it
+		if ca.GetName() == rootClusterName || (wantLeaf && ca.GetName() == cf.LeafClusterName) {
+			// if we've found the requested leaf cluster, mark that we've found it
+			if ca.GetName() == cf.LeafClusterName {
+				leafExists = true
+			}
+			for i, key := range ca.GetTrustedSSHKeyPairs() {
+				log.Debugf("%v CA [%v]: %v", ca.GetName(), i, key)
+				kh, err := sshutils.MarshalKnownHost(sshutils.KnownHost{
+					Hostname:      ca.GetClusterName(),
+					AuthorizedKey: key.PublicKey,
+				})
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				_, _, hostCABytes, _, _, err := ssh.ParseKnownHosts([]byte(kh))
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				hostCAPublicKeys[ca.GetName()] = append(hostCAPublicKeys[ca.GetName()], strings.TrimPrefix(strings.TrimSpace(string(ssh.MarshalAuthorizedKey(hostCABytes))), constants.SSHRSAType+" "))
```

Comment:
kinda a nitpick, but for readability's sake it might be nice to compute the new value on its own line (maybe even 2)

Reply:
Agree, it's messy as-is. Two lines is an improvement.